src/agreement.rs

```diff
@@ -1,42 +1,241 @@
 //! Binary Byzantine agreement protocol from a common coin protocol.
 
+use std::collections::{BTreeMap, BTreeSet, HashMap, VecDeque};
+use std::hash::Hash;
+
 use proto::AgreementMessage;
-use std::collections::{BTreeSet, VecDeque};
 
-#[derive(Default)]
-pub struct Agreement {
+pub struct Agreement<NodeUid> {
+    /// The UID of the corresponding node.
+    uid: NodeUid,
+    num_nodes: usize,
+    num_faulty_nodes: usize,
+    epoch: u32,
     input: Option<bool>,
-    _bin_values: BTreeSet<bool>,
+    /// Bin values. Reset on every epoch update.
+    bin_values: BTreeSet<bool>,
+    /// Values received in BVAL messages. Reset on every epoch update.
+    received_bval: HashMap<NodeUid, BTreeSet<bool>>,
+    /// Sent BVAL values. Reset on every epoch update.
+    sent_bval: BTreeSet<bool>,
+    /// Values received in AUX messages. Reset on every epoch update.
+    received_aux: HashMap<NodeUid, BTreeSet<bool>>,
+    /// All the output values in all epochs.
+    outputs: BTreeMap<u32, bool>,
```

Comment:
Agreement can output multiple times. That's why the termination criterion is more complex than whether the instance has output a value. I have to correct the code to reflect that. At the moment only one value is stored per agreement instance.

Reply:
Right, that's a bit ambiguous in the paper. I think the reason for the complex termination criterion is just that a node needs to keep sending messages to help the other nodes output, too. The formulation "then output b" should really be read as "then output b, unless you already have done so". For the user of the module it will certainly be more convenient if it only outputs once?